godot-core/src/engine/translate.rs

```diff
@@ -0,0 +1,104 @@
+///  methods.
+///
+/// Takes a format string literal, with optional arguments. Optionally, `context` for potentially ambiguous words can be
+/// added before the format arguments, separated with a `;`.
+///
+/// For maximum flexibility and minimum ambiguity, prefer inlined, positional, or named arguments in the format strings:
+/// ```no_run
+/// # #[macro_use] extern crate godot;
+/// # use godot::builtin::Vector2i;
+/// # let a = Vector2i { x: 0, y: 0 };
+/// # let b = Vector2i { x: 0, y: 0 };
+/// # let context = "context";
+/// use godot::engine::translate::tr;
+///
+/// // Good
+/// tr!(context; "{a} is a {b}"); // inlined, with context
+/// tr!("{0} is a {1}", a, b); // positional, without context
+/// tr!("{c} is a {d}", c = a.x, d = b.y); // named (inlining not possible here)
+///
+/// // Not so good, much more fragile
+/// // and does not easily allow ordering arguments differently in other translations.
+/// tr!("{} is a {}", a, b);
+/// ```
+/// The methods are called from the [`Engine`](crate::engine::Engine) singleton.
+///
+/// see also: <https://docs.godotengine.org/en/stable/tutorials/i18n/internationalizing_games.html#translation-contexts>
+#[macro_export]
+macro_rules! tr {
+    ($fmt:literal $(, $($args:tt)*)?) => {
+        $crate::engine::Engine::singleton()
+            .tr(format!($fmt $(, $($args)*)?).into())
+    };
+
+    ($context:expr; $fmt:literal $(, $($args:tt)*)?) => {
+        $crate::engine::Engine::singleton()
+            .tr_ex(format!($fmt $(, $($args)*)?).into())
+            .context(format!("{}", $context).into())
+            .done()
+    };
+}
+
+/// A convenience macro for using the [`Object::tr_n()`](crate::engine::Object::tr_n()) and
+/// [`Object::tr_n_ex()`](crate::engine::Object::tr_n_ex()) methods.
+///
+/// Allows for the use of format strings with arbitrary arguments. `n` is given prior to the format string, followed by `;`.
+/// Optionally, `context` for potentially ambiguous words can be added with `,` after `n` and before `;`.
+///
+/// For maximum flexibility and minimum ambiguity, prefer inlined, positional, or named arguments in the format strings:
+/// ```no_run
+/// # #[macro_use] extern crate godot;
+/// # use godot::builtin::Vector2i;
+/// # let a = Vector2i { x: 0, y: 0 };
+/// # let b = Vector2i { x: 0, y: 0 };
+/// # let context = "context";
+/// # let n = 2;
+/// use godot::engine::translate::tr_n;
+///
+/// // Good
+/// tr_n!(n, context; "{a} is a {b}", "{a}s are {b}s"); // inlined, with context
+/// tr_n!(n; "{0} is a {1}", "{0}s are {1}s", a, b); // positional, without context
+/// tr_n!(n; "{c} is a {d}", "{c}s are {d}s", c = a.x, d = b.y); // named (inlining not possible here)
+///
+/// // Not so good, much more fragile
+/// // and does not easily allow ordering arguments differently in other translations.
+/// // Additionally, such syntax requires that BOTH format strings must use ALL arguments
```

Comment:
I still have a question about
> and does not easily allow ordering arguments differently in other translations.

What do you mean exactly? The format args are not passed to other translations/languages, they are resolved in `format!` in the original text.

So I would either clarify this or leave it out; right now it's a bit confusing.

Also, as mentioned take care of proper sentence structure (full stop). Thanks!

Reply:
As currently written, that's true, they're pre-formatted. However, [this](https://docs.godotengine.org/en/stable/tutorials/i18n/internationalizing_games.html#placeholders) was my reference. The `tr` call in gdscript is with the format literal alone, so it will grab the appropriate language string (presumably with those same format specifiers) and incorporate the arguments into *that* as the final result.

I admit that I didn't realize that at first, but since it's a macro, the underlying implementation could be expanded to allow for that, in which case it's good advice. It should actually probably work that way already, which is an oversight on my part.

Reply:
> Actually, allowing for this functionality is non-trivial, as `format!` and related macros require a compile-time known string literal

Yes, that's why I was wondering how the user could substitute/reorder anything 🙂 



> I think having the syntax support it at the very least means it can be added.

That's fair, but the documentation is currently wrong and hints at a possibility that isn't there. I would thus omit this part; we can always re-add it, if we add the feature. To already encourage good habits, maybe something like "using named or ordered placeholders instead of `{}` may make dynamic arguments easier, once gdext supports them". Feel free to find a better wording or omit entirely 😀